svf-llvm/tools/CFL/CMakeLists.txt

```diff
@@ -3,7 +3,13 @@ if(DEFINED IN_SOURCE_BUILD)
 else()
   add_executable(cfl cfl.cpp)
 
-  target_link_libraries(cfl SvfLLVM ${llvm_libs})
+  separate_arguments(LLVM_DEFINITIONS_LIST NATIVE_COMMAND ${LLVM_DEFINITIONS})
+  target_include_directories(cfl PUBLIC include)
+  target_include_directories(cfl SYSTEM PUBLIC ${LLVM_INCLUDE_DIRS})
+  target_link_directories(cfl PUBLIC ${LLVM_LIBRARY_DIRS})
+  target_compile_definitions(cfl PUBLIC ${LLVM_DEFINITIONS})
+
+  target_link_libraries(cfl SvfCore SvfLLVM)
```

Comment:
It looks to me like every tool has these repetitive lines. Could we move them to parent cmakelist?

Reply:
Of course; I did this because I thought the linking was going wrong but on my system this is difficult to reproduce as my LLVM instance's library directory is also in `$LD_LIBRARY_PATH`, so I cannot reproduce the issue with finding `libLLVM.[so|a]`. Especially since the CMake output during the git workflow don't include the build command itself, so I can't see if the issue is due to the `rpath` not being set right, or if somehow `link_directories()` is not propagating to targets defined in subdirectories.

Either way, it is not necessary to explicitly link against the LLVM library anyway since the library is defined using `add_llvm_library`, so I removed this part. Moreover, since SVF aims to support building in- and out-of-tree, the SVF-LLVM tools can/should also be defined using `add_llvm_executable` or even `add_llvm_tool` (I didn't do the latter for now as that means they only get built when specific build flags are set for LLVM). This simplifies the linking & the makefiles a bit since there is no longer a need for in-tree detection and what not.

Note that I've also modified `build.sh` and `setup.sh` to not only append LLVM's binary directory (for local installations by the build script) to the `$PATH` variable, but also its library directories to `$LD_LIBRARY_PATH` so that the linker can reliable find the libraries (idem ditto for Z3). I also did the same for the library files produced by SVF so targets linking against SVF can find them easier. Please let me know if you think this should be done differently!